libflux/src/semantic/tests.rs

```diff
@@ -255,715 +255,2377 @@ fn literals() {
 // variables that are type variables.
 
 #[test]
-fn string_expr() {
-    let src = r#"message = "Hello, ${name}!""#;
-    //    test_infer! {
-    //        env: map![
-    //            "name" => "forall [] t0",
-    //        ],
-    //        src: src,
-    //        exp: map![
-    //            "name" => "forall [] string",
-    //            "message" => "forall [] string",
-    //        ],
-    //    }
+fn string_interpolation() {
+    // test_infer! {
+    //     env: map![
+    //         "name" => "forall [] t0",
+    //     ],
+    //     src: src,
+    //     exp: map![
+    //         "name" => "forall [] string",
+    //         "message" => "forall [] string",
+    //     ],
+    // }
+    test_infer! {
+        env: map![
+            "name" => "forall [] string",
+        ],
+        src: r#"
+            message = "Hello, ${name}!"
+        "#,
+        exp: map![
+            "message" => "forall [] string",
+        ],
+    }
+    test_infer_err! {
+        env: map![
+            "name" => "forall [] bool",
+        ],
+        src: r#"
+            "Hello, ${name}!"
+        "#,
+    }
     test_infer_err! {
         env: map![
             "name" => "forall [] int",
         ],
-        src: src,
+        src: r#"
+            "Hello, ${name}!"
+        "#,
     }
-    test_infer! {
+    test_infer_err! {
         env: map![
-            "name" => "forall [] string",
+            "name" => "forall [] uint",
         ],
-        src: src,
-        exp: map![
-            "message" => "forall [] string",
+        src: r#"
+            "Hello, ${name}!"
+        "#,
+    }
+    test_infer_err! {
+        env: map![
+            "name" => "forall [] float",
+        ],
+        src: r#"
+            "Hello, ${name}!"
+        "#,
+    }
+    test_infer_err! {
+        env: map![
+            "name" => "forall [] duration",
+        ],
+        src: r#"
+            "Hello, ${name}!"
+        "#,
+    }
+    test_infer_err! {
+        env: map![
+            "name" => "forall [] time",
+        ],
+        src: r#"
+            "Hello, ${name}!"
+        "#,
+    }
+    test_infer_err! {
+        env: map![
+            "name" => "forall [] regexp",
+        ],
+        src: r#"
+            "Hello, ${name}!"
+        "#,
+    }
+    test_infer_err! {
+        env: map![
+            "name" => "forall [] [int]",
+        ],
+        src: r#"
+            "Hello, ${name}!"
+        "#,
+    }
+    test_infer_err! {
+        env: map![
+            "name" => "forall [] {a: int | b: float}",
+        ],
+        src: r#"
+            "Hello, ${name}!"
+        "#,
+    }
+    test_infer_err! {
+        env: map![
+            "name" => "forall [t0] (x: t0) -> t0",
         ],
+        src: r#"
+            "Hello, ${name}!"
+        "#,
```

Comment:
So, no type except from `string` is `Stringable`?

Do we require the user to manually `string(v: ...)`?

Reply:
Right now yes, but we will add a `Stringable` kind to the type system eventually so that one can interpolate other expression types into string expressions.